apps/aecore/lib/aecore/contract/contract_state_tree.ex

```diff
@@ -50,30 +44,37 @@ defmodule Aecore.Contract.ContractStateTree do
     update_store(store_id, old_contract_store, store, updated_contract_tree)
   end
 
-  @spec get_contract(contracts_state(), binary()) :: Contract.t()
-  def get_contract(contract_tree, key) do
-    case PatriciaMerkleTree.lookup(contract_tree, key) do
-      {:ok, serialized} ->
-        {:ok, deserialized} = Serialization.rlp_decode_anything(serialized)
-
-        identified_id = Identifier.create_identity(key, :contract)
+  @spec process_struct(Contract.t(), binary(), contracts_state()) :: Contract.t()
+  def process_struct(%Contract{} = deserialized_value, key, tree) do
+    identified_id = Identifier.create_identity(key, :contract)
+    store_id = Contract.store_id(%{deserialized_value | id: identified_id})
 
-        store_id = Contract.store_id(%{deserialized | id: identified_id})
-
-        %Contract{
-          deserialized
-          | id: identified_id,
-            store: get_store(store_id, contract_tree)
-        }
+    %Contract{
+      deserialized_value
+      | id: identified_id,
+        store: get_store(store_id, tree)
+    }
+  end
 
-      _ ->
-        :none
-    end
+  def process_struct(deserialized_value, _key, _tree) do
+    {:error,
+     "#{__MODULE__}: Invalid data type: #{deserialized_value.__struct__} but expected %Contract{}"}
   end
 
-  @spec root_hash(contracts_state()) :: hash()
-  def root_hash(contract_tree) do
-    PatriciaMerkleTree.root_hash(contract_tree)
+  defp get_store(store_id, tree) do
+    keys = PatriciaMerkleTree.all_keys(tree)
+    store_id_bit_size = (@contract_key_size + 1) * 8
+
+    Enum.reduce(keys, %{}, fn key, store_acc ->
```

Comment:
@DanielaIvanova
We can have a lot of contracts on the blockchain. Ethereum currently has more than a million of then (https://etherscan.io/accounts/c).

This code has a time complexity of O(N) which means that this code will be unusable in production -> Each O(log(N)) lookup will degrade to O(N).

The operation implemented here corresponds in fact to taking an subset of the whole trie which can be done in O(logN). Open an issue to implement this operation in our MerklePatriciaLibrary - we will need to implement a module providing a read/write view onto a subset of a Merkle Patricia Trie - Please note that we cannot simply lookup all the keys in a sub tree and create a new map from this as this could be potentially time consuming. 

@cytadela8 @thepiwo I would suggest that we merge this PR to master but open an issue about this so we do not forget to rewrite this. 

@DanielaIvanova Add a comment to the code stating that this snipped has a time complexity od O(n) and needs to be rewritten.

Reply:
This code is already in the master. I will add the comment . This code was written @cheezus1

Reply:
@thepiwo I opened the issue - https://github.com/aeternity/elixir-node/issues/668